docs/source/tutorial_metrics.md

```diff
@@ -0,0 +1,372 @@
+you can do this, your code could not be run in [_distributed
+mode_](tutorial_fast). If we only returned a single float, we would not be able
+to know if some distributed workers received more or fewer examples than
+others. However, when we explicitly store the numerator and denominator, we can
+combine and reduce the across multiple nodes, enabling us to train on hundreds
+of GPUs, while still ensuring correctness in all our metrics.
+
+In addition to AverageMetric, there is also
+[SumMetric](parlai.core.metrics.SumMetric), which keeps a running
+sum. SumMetric and AverageMetric are the most common ways to construct custom
+metrics, but others exist as well. For a full list (and views into advanced
+cases), please see the [metrics API documentation](metrics_api).
+
+## Agent (model) level metrics
+
+In the above example, we worked on a metric defined by a Teacher. However,
+sometimes our models will have special metrics that only they want to compute,
+which we call an Agent-level metric. Perplexity is one example.
+
+To compute model-level metrics, we can define either a global metric, or a
+local metric. Global metrics can be computed anywhere, and are easy to use,
+but cannot distinguish between different teachers when multitasking. We'll
+look at another example, counting the number of times the teacher says "hello".
+
+### Global metrics
+
+A global metric is computed anywhere in the model, and has an
+interface similar to that of the teacher:
+
+```python
+agent.global_metrics.add('my_metric', AverageMetric(1, 2))
+```
+
+Global metrics are called as such because they can be called anywhere in agent
+code. For example, we can add a metric that counts the number of times the
+model sees the word "hello" in `observe`. We'll do this while extending
+the `TransformerGeneratorAgent`, so that we can combined it with the BlenderBot
+model we used earlier.
+
+```python
+from parlai.core.metrics import AverageMetric
+from parlai.core.loader import register_agent
+from parlai.agents.transformer.transformer import TransformerGeneratorAgent
+
+
+@register_agent('GlobalHelloCounter')
+class GlobalHelloCounterAgent(TransformerGeneratorAgent):
+    def observe(self, observation):
+        retval = super().observe(observation)
+        if 'text' in observation:
+            text = observation['text']
+            self.global_metrics.add(
+                'global_hello', AverageMetric(int('hello' in text), 1)
+            )
+        return retval
+
+
+if __name__ == '__main__':
+    from parlai.scripts.eval_model import EvalModel
+
+    EvalModel.main(
+        task='dailydialog',
+        model='GlobalHelloCounter',
+        model_file='zoo:blender/blender_90M/model',
+        batchsize=32,
+    )
+```
+
+Running the script, we see that our new metric appears. Note that it is
+different than the metric in the first half of the tutorial: that is because
+previously we were counting the number of times the model said hello (a lot),
```

Comment:
Previously, we had a teacher agent and metric was calculated on its actions. If I am not wrong, those are coming from dataset. I don't understand why it says here that the metric counts the number of times the model says hello then.

Reply:
Those were counting how many 'text' in model_response which model_response is The full model response, including any extra fields the model may have sent. (see L135) What I believe the difference is the previous one is counting how frequent the model says hello  given that it sees the dataset, and here it counts how often the model observe "hello" the context. I could be a little confusing as the naming of the metrics could suggest otherwise.

Reply:
Thanks for flagging, I agree. The confusion is worth making unambiguous

Reply:
I've added some clarifications and extra tip section emphasizing this.